test/gtest/uct/test_peer_failure.cc

```diff
@@ -181,6 +181,33 @@ class test_uct_peer_failure : public uct_test {
                (is_time_out ? UCS_ERR_TIMED_OUT : UCS_ERR_OUT_OF_RANGE);
     }
 
+    ucs_status_t add_pending(uct_ep_h ep, pending_send_request_t &req) {
+        req.ep       = ep;
+        req.uct.func = pending_cb;
+        return uct_ep_pending_add(ep, &req.uct, 0);
+    }
+
+    void fill_resources(bool expect_error, ucs_time_t loop_end_limit) {
+        const uint64_t send_data = 0;
+        ucs_status_t status;
+        do {
+            status = uct_ep_am_short(ep0(), 0, 0, &send_data,
+                                     sizeof(send_data));
+        } while ((status == UCS_OK) && (ucs_get_time() < loop_end_limit));
+
+        if (status == UCS_OK) {
+            UCS_TEST_SKIP_R("unable to fill the UCT resources");
+        } else if (status != UCS_ERR_NO_RESOURCE) {
+            if (expect_error) {
```

Comment:
but it could be `UCS_ERR_IO_ERROR` in case of TCP

Reply:
why IO_ERR and not ENDPOINT_TIMEOUT?
also, AFAIK TCP does not even support err handling now

Reply:
if TCP detects an error during doing `send()`, it returns IO_ERROR and calls base's `set_ep_failed()` that sets all function pointers to return ENDPOIN_TIMEOUT
yes, TCP doesn't support, but this is idea in #5294. do you think it's not correct?

Reply:
IB TLs could detect error in progress() while TCP is able to detect it in AM/PUT operations

Reply:
TCP should return some error meaning code if there is send() failure, which will hold true for UCS_IS_ENDPOINT_ERROR

Reply:
done

Reply:
I don't think that i would work, since if even UD does sends after peer failure detection, send will return `UCS_IS_ENDPOINT_ERROR` that is false from `UCS_IS_ENDPOINT_ERROR()`

Reply:
which error will it return? can you point to relevant UD code?

Reply:
> which error will it return? can you point to relevant UD code?

I meant this is common code for all UCTs:
https://github.com/openucx/ucx/blob/6f2ee46e81da7db7d73936dd0929756e23709412/src/uct/base/uct_iface.c#L349

Reply:
right, ucs_empty_function_return_ep_timeout() return value would be "true" for `UCS_IS_ENDPOINT_ERROR()`

Reply:
```
#define UCS_IS_ENDPOINT_ERROR(_code) \
    (((_code) <= UCS_ERR_FIRST_ENDPOINT_FAILURE) && \
     ((_code) >= UCS_ERR_LAST_ENDPOINT_FAILURE))
```
where `UCS_ERR_FIRST_ENDPOINT_FAILURE = -60` and `UCS_ERR_LAST_ENDPOINT_FAILURE = -79`
but `UCS_ERR_ENDPOINT_TIMEOUT = -80`
so, status has to be `status <= -60` and `status >= -79` (`status = [-79; -60]`)
should we fix `UCS_ERR_ENDPOINT_TIMEOUT` to be `-79 <= x <= -60`, but I guess it is not backward compatible change

Reply:
> should we fix `UCS_ERR_ENDPOINT_TIMEOUT` to be `-79 <= x <= -60`, but I guess it is not ABI compatible change

Can we fix UCS_ERR_LAST_ENDPOINT_FAILURE to be -89 for example?

Reply:
> Can we fix UCS_ERR_LAST_ENDPOINT_FAILURE to be -89 for example?

yes, it's better

Reply:
@yosefe done